Migrate rerunning-dags guide (#1186)
astronomer/docs: guides/rerunning-dags.md

---
title: "Rerun Airflow DAGs"

> sidebar_label: "Rerun Airflow DAGs"

Reviewer: ```suggestion
sidebar_label: "Rerun DAGs"
```

> description: "How to use catchup, backfill, and cleared task instances in Airflow."

Reviewer: ```suggestion
description: "How to catchup, backfill, and clear task instances in Airflow."
```

id: rerunning-dags
---

Running DAGs whenever you want is one of the most powerful and flexible features of Airflow. Scheduling DAGs can ensure future DAG runs happen at the right time, but you also have options for running DAGs in the past. For example, you might need to run a DAG in the past if:

- You need to rerun a failed task for one or multiple DAG runs.
- You want to deploy a DAG with a start date of one year ago and trigger all DAG runs that would have been scheduled in the past year.
- You have a running DAG and realize you need it to process data for two months prior to the DAG's start date.

> In this guide, you'll learn how to rerun tasks or DAGs and trigger historical DAG runs, and review the Airflow concepts of catchup and backfill. If you're looking for additional information about DAG scheduling, see [Scheduling and Timetables in Airflow](https://www.astronomer.io/guides/scheduling-in-airflow) guide.

Reviewer: ```suggestion
In this guide, you'll learn how to rerun tasks or DAGs and trigger historical DAG runs, and review the Airflow concepts of catchup and backfill. If you're looking for additional information about DAG scheduling, see [Scheduling and timetables in Airflow](https://www.astronomer.io/guides/scheduling-in-airflow).
```

> ## Rerunning tasks

Reviewer: ```suggestion
## Rerun tasks
```

- Failed: Clears only failed instances of any task instances selected based on the above options.

After you make your selections and click **Clear**, the Airflow UI shows a summary of the task instances that will be cleared and asks you to confirm. Use this to ensure your selections are applied as intended.

![Task Instance Summary](/img/guides/task_instance_confirmation.png)

You can also run the following command in the Airflow CLI to clear task statuses:

``` bash
airflow tasks clear [-h] [-R] [-d] [-e END_DATE] [-X] [-x] [-f] [-r]
                    [-s START_DATE] [-S SUBDIR] [-t TASK_REGEX] [-u] [-y]
                    dag_id
```

For more info on the positional arguments for the `clear` command, see the [Airflow documentation](https://airflow.apache.org/docs/apache-airflow/stable/cli-and-env-variables-ref.html#clear).

Clearing or changing task statuses directly in the Airflow metastore can cause unexpected behavior in Airflow.

To clear a full DAG run, go to the Tree View in the Airflow UI and then click **Clear** as shown in the following image. 

![Clear DAG Status](/img/guides/clear_dag_ui.png)

> ### Clearing all tasks

Reviewer: ```suggestion
### Clear all tasks
```

```bash
airflow dags backfill [-h] [-c CONF] [--delay-on-limit DELAY_ON_LIMIT] [-x]
                      [-n] [-e END_DATE] [-i] [-I] [-l] [-m] [--pool POOL]
                      [--rerun-failed-tasks] [--reset-dagruns] [-B]
                      [-s START_DATE] [-S SUBDIR] [-t TASK_REGEX] [-v] [-y]
                      dag_id
```

For example, `airflow dags backfill -s 2021-11-01 -e 2021-11-02 example_dag` backfills `example_dag` from November 1st-2nd 2021. For more information about the backfill parameter, see the [Airflow documentation](https://airflow.apache.org/docs/apache-airflow/stable/cli-and-env-variables-ref.html#backfill). 

When using backfill keep the following considerations in mind:

- Consider your available resources. If your backfill will trigger many DAG runs, you might want to add some of the catchup parameters to your DAG.
- Clearing the task or DAG status of a backfilled DAG run does not rerun the task or DAG.

> If you don't have access to the Airflow CLI, deploy a copy of the DAG with a new name and a start date that is the date you want to backfill to. Airflow will consider this a separate DAG so you won't see all the DAG runs and task instances in the same place, but it would accomplish running the DAG for data in the desired time period. If you have a small number of DAG runs to backfill, you can trigger them manually from the Airflow UI and choose the desired logical date as shown in the following image:

Reviewer: ```suggestion
Alternatively, you can deploy a copy of the DAG with a new name and a start date that is the date you want to backfill to. Airflow will consider this a separate DAG so you won't see all the DAG runs and task instances in the same place, but it would accomplish running the DAG for data in the desired time period. If you have a small number of DAG runs to backfill, you can trigger them manually from the Airflow UI and choose the desired logical date as shown in the following image:
```
